<h2 align='center' style='color:blue'>Dropout Regularization In Deep Neural Network</h2>

This is a dataset that describes sonar chirp returns bouncing off different services. The 60 input variables are the strength of the returns at different angles. It is a **binary classification problem** that requires a model to differentiate rocks from metal cylinders.

Dataset information: https://archive.ics.uci.edu/ml/datasets/Connectionist+Bench+(Sonar,+Mines+vs.+Rocks)
Download it from here: https://archive.ics.uci.edu/ml/machine-learning-databases/undocumented/connectionist-bench/sonar/sonar.all-data

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
import warnings
warnings.filterwarnings('ignore')

In [3]:
df = pd.read_csv("/Users/athulnambiar/Desktop/PROJECTS/DL-YT/DROPOUT ROCK MINE PREDICTION/sonar_dataset.csv", header=None)
df.sample(5)

0       1       2       3       4       5       6       7       8   \
20   0.0473  0.0509  0.0819  0.1252  0.1783  0.3070  0.3008  0.2362  0.3830   
194  0.0392  0.0108  0.0267  0.0257  0.0410  0.0491  0.1053  0.1690  0.2105   
37   0.0333  0.0221  0.0270  0.0481  0.0679  0.0981  0.0843  0.1172  0.0759   
73   0.0139  0.0222  0.0089  0.0108  0.0215  0.0136  0.0659  0.0954  0.0786   
157  0.0201  0.0178  0.0274  0.0232  0.0724  0.0833  0.1232  0.1298  0.2085   

         9   ...      51      52      53      54      55      56      57  \
20   0.3759  ...  0.0193  0.0118  0.0064  0.0042  0.0054  0.0049  0.0082   
194  0.2471  ...  0.0083  0.0080  0.0026  0.0079  0.0042  0.0071  0.0044   
37   0.0920  ...  0.0022  0.0032  0.0060  0.0054  0.0063  0.0143  0.0132   
73   0.1015  ...  0.0062  0.0072  0.0113  0.0012  0.0022  0.0025  0.0059   
157  0.2720  ...  0.0131  0.0049  0.0104  0.0102  0.0092  0.0083  0.0020   

         58      59  60  
20   0.0028  0.0027   R  
194  0.0022  0.0014   M  
37   0.0051  0.0041   R  
73   0.0039  0.0048   R  
157  0.0048  0.0036   M  

[5 rows x 61 columns]

In [4]:
df.shape

(208, 61)

In [5]:
# check for nan values
df.isna().sum()

0     0
1     0
2     0
3     0
4     0
     ..
56    0
57    0
58    0
59    0
60    0
Length: 61, dtype: int64

In [6]:
df.columns

Index([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35,
       36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53,
       54, 55, 56, 57, 58, 59, 60],
      dtype='int64')

In [7]:
df[60].value_counts() # label is not skewed

60
M    111
R     97
Name: count, dtype: int64

In [8]:
X = df.drop(60, axis=1)
y = df[60]
y.head()

0    R
1    R
2    R
3    R
4    R
Name: 60, dtype: object

In [9]:
y = pd.get_dummies(y, drop_first=True)
y.sample(5) # R --> 1 and M --> 0

R
154  False
168  False
89    True
114  False
148  False

In [10]:
y.value_counts()

R    
False    111
True      97
Name: count, dtype: int64

In [11]:
X.head()

0       1       2       3       4       5       6       7       8   \
0  0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1  0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2  0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3  0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4  0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   

       9   ...      50      51      52      53      54      55      56  \
0  0.2111  ...  0.0232  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180   
1  0.2872  ...  0.0125  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140   
2  0.6194  ...  0.0033  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316   
3  0.1264  ...  0.0241  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050   
4  0.4459  ...  0.0156  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072   

       57      58      59  
0  0.0084  0.0090  0.0032  
1  0.0049  0.0052  0.0044  
2  0.0164  0.0095  0.0078  
3  0.0044  0.0040  0.0117  
4  0.0048  0.0107  0.0094  

[5 rows x 60 columns]

In [12]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1)

In [13]:
X_train.head()

0       1       2       3       4       5       6       7       8   \
67   0.0368  0.0403  0.0317  0.0293  0.0820  0.1342  0.1161  0.0663  0.0155   
14   0.0124  0.0433  0.0604  0.0449  0.0597  0.0355  0.0531  0.0343  0.1052   
164  0.0163  0.0198  0.0202  0.0386  0.0752  0.1444  0.1487  0.1484  0.2442   
179  0.0394  0.0420  0.0446  0.0551  0.0597  0.1416  0.0956  0.0802  0.1618   
19   0.0126  0.0149  0.0641  0.1732  0.2565  0.2559  0.2947  0.4110  0.4983   

         9   ...      50      51      52      53      54      55      56  \
67   0.0506  ...  0.0058  0.0091  0.0160  0.0160  0.0081  0.0070  0.0135   
14   0.2120  ...  0.0078  0.0083  0.0057  0.0174  0.0188  0.0054  0.0114   
164  0.2822  ...  0.0027  0.0077  0.0026  0.0031  0.0083  0.0020  0.0084   
179  0.2558  ...  0.0118  0.0146  0.0040  0.0114  0.0032  0.0062  0.0101   
19   0.5920  ...  0.0153  0.0092  0.0035  0.0098  0.0121  0.0006  0.0181   

         57      58      59  
67   0.0067  0.0078  0.0068  
14   0.0196  0.0147  0.0062  
164  0.0108  0.0083  0.0033  
179  0.0068  0.0053  0.0087  
19   0.0094  0.0116  0.0063  

[5 rows x 60 columns]

# Using Deep Learning Model

## Model without Dropout Layer

In [14]:
import tensorflow as tf
from tensorflow import keras

In [15]:
model = keras.Sequential([
    keras.layers.Dense(60, input_dim=60, activation='relu'),
    keras.layers.Dense(30, activation='relu'),
    keras.layers.Dense(15, activation='relu'),
    keras.layers.Dense(1, activation='sigmoid')
])

model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

model.fit(X_train, y_train, epochs=100, batch_size=8)

Epoch 1/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 958us/step - accuracy: 0.5709 - loss: 0.6762
Epoch 2/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 757us/step - accuracy: 0.6556 - loss: 0.6542
Epoch 3/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 708us/step - accuracy: 0.6291 - loss: 0.6348
Epoch 4/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 781us/step - accuracy: 0.7495 - loss: 0.5983
Epoch 5/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 704us/step - accuracy: 0.6776 - loss: 0.5901
Epoch 6/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 657us/step - accuracy: 0.7640 - loss: 0.5397
Epoch 7/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 673us/step - accuracy: 0.7231 - loss: 0.5134
Epoch 8/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 576us/step - accuracy: 0.7780 - loss: 0.4873
Epoch 9/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 526us/step - accuracy: 0.7628 - loss: 0.4686
Epoch 10/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 517us/step - accuracy: 0.7655 - loss: 0.4869
Epoch 11/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.8348 - loss: 0.4143 
Epoch 12/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s

In [16]:
model.evaluate(X_test, y_test)

2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.7292 - loss: 1.2931 


[1.0927209854125977, 0.75]

Training Accuracy >>> Test Accuracy

In [17]:
y_pred = model.predict(X_test).reshape(-1)
print(y_pred[:10])

# round the values to nearest integer ie 0 or 1
y_pred = np.round(y_pred)
print(y_pred[:10])

2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 19ms/step
[8.6753484e-09 9.8588353e-01 9.9556965e-01 6.9842186e-05 1.0000000e+00
 9.9999321e-01 2.7193258e-02 1.0000000e+00 4.9374881e-05 1.0000000e+00]
[0. 1. 1. 0. 1. 1. 0. 1. 0. 1.]


In [18]:
y_test[:10]

R
186  False
155  False
165  False
200  False
58    True
34    True
151  False
18    True
202  False
62    True

In [19]:
from sklearn.metrics import confusion_matrix , classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

       False       0.72      0.85      0.78        27
        True       0.80      0.64      0.71        25

    accuracy                           0.75        52
   macro avg       0.76      0.75      0.75        52
weighted avg       0.76      0.75      0.75        52



### Model with Dropout Layer

In [20]:
modeld = keras.Sequential([
    keras.layers.Dense(60, input_dim=60, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(30, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(15, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(1, activation='sigmoid')
])

modeld.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

modeld.fit(X_train, y_train, epochs=100, batch_size=8)

Epoch 1/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 1s 764us/step - accuracy: 0.5160 - loss: 0.7075
Epoch 2/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 663us/step - accuracy: 0.4756 - loss: 0.7365
Epoch 3/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 662us/step - accuracy: 0.4869 - loss: 0.7193
Epoch 4/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 640us/step - accuracy: 0.3971 - loss: 0.7356
Epoch 5/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 610us/step - accuracy: 0.5060 - loss: 0.6851
Epoch 6/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 615us/step - accuracy: 0.5310 - loss: 0.6903
Epoch 7/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.5094 - loss: 0.7165 
Epoch 8/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 943us/step - accuracy: 0.5919 - loss: 0.6702
Epoch 9/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 607us/step - accuracy: 0.5605 - loss: 0.6922
Epoch 10/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 542us/step - accuracy: 0.4725 - loss: 0.7199
Epoch 11/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.5596 - loss: 0.6898 
Epoch 12/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 

In [21]:
modeld.evaluate(X_test, y_test)

2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.7316 - loss: 0.4757 


[0.4118070900440216, 0.7692307829856873]

Training Accuracy is still good but Test Accuracy Improved

In [22]:
y_pred = modeld.predict(X_test).reshape(-1)
print(y_pred[:10])

# round the values to nearest integer ie 0 or 1
y_pred = np.round(y_pred)
print(y_pred[:10])

2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 18ms/step
[0.00188628 0.94028705 0.8463954  0.08041368 0.99984336 0.9764812
 0.89269763 0.9999723  0.05277117 0.9999451 ]
[0. 1. 1. 0. 1. 1. 1. 1. 0. 1.]


In [23]:
from sklearn.metrics import confusion_matrix , classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

       False       0.76      0.81      0.79        27
        True       0.78      0.72      0.75        25

    accuracy                           0.77        52
   macro avg       0.77      0.77      0.77        52
weighted avg       0.77      0.77      0.77        52



**You can see that by using dropout layer test accuracy increased from 0.77 to 0.81**